New: Add Peierls exact formulation

bobmyhill opened the pull request:
This PR adds the exact formulation for Peierls creep to the Peierls creep rheology module, along with the associated strain rate and derivative functions.
It also adds a new test using the analytical calculations by @mibillen and @naliboff.

Based on #3723.

bobmyhill commented:
Hi @naliboff, I think this can be checked over now that your PR is merged :)

naliboff commented on source/material_model/rheology/peierls_creep.cc:
```diff
@@ -107,15 +199,124 @@ namespace aspect
                            "Select what type of Peierls creep flow law to use. Currently, the "
-                           "only available option is an approximation to Peierls creep, which "
-                           "uses the strain rate invariant, rather than stress, as an input. ");
+                           "available options are 'exact', which uses a Newton-Raphson iteration "
```
maybe `Newton-Raphson iterative method`?

naliboff commented on source/material_model/rheology/peierls_creep.cc:
```diff
@@ -107,15 +199,124 @@ namespace aspect
+
+        // Viscosity iteration parameters
+        prm.declare_entry ("Peierls strain rate residual tolerance", "1e-22", Patterns::Double(0.),
+                           "Tolerance for correct Peierls creep strain rate.");
```
```suggestion
                           "Tolerance for the iterative solve to find the correct Peierls creep strain rate.");
```

naliboff commented on include/aspect/material_model/rheology/peierls_creep.h:
```diff
@@ -71,11 +115,43 @@ namespace aspect
+
+          /**
+           * Compute the strain rate and first stress derivative
+           * as a function of stress based on the selected Peierls creep law.
```
At first reading I was a bit confused about the difference between this function and the two above, but this became immediately clear after reading on. Perhaps worth adding a note that this function utilizes the two functions above?  If this seems excessive feel free to disregard.

naliboff commented on source/material_model/rheology/peierls_creep.cc:
```diff
@@ -38,61 +38,153 @@ namespace aspect
+         * A generalised Peierls creep formulation. The Peierls creep expression
+         * for the strain rate has multiple stress-dependent terms, and cannot be
+         * directly inverted to find an expression for viscosity in terms of
+         * strain rate. For this reason, a Newton Raphson iteration is required,
```
```suggestion
         * strain rate. For this reason, a Newton-Raphson iteration is required,
```

naliboff commented on source/material_model/rheology/peierls_creep.cc:
```diff
@@ -38,61 +38,153 @@ namespace aspect
+
+        double strain_rate_deriv = 0;
+        unsigned int stress_iteration = 0;
+        while (std::abs(strain_rate_residual) > strain_rate_residual_threshold
```
In diffusion_dislocation a fixpoint iteration is used if the Newton iteration does not succeed. Are we likely to run into the same issues here? If so, should we add this option in now? I have no practical experience with this type of rheology, and I will defer to your judgement. Worst case scenario, we add in the fixpoint iterative scheme if we find the Newton method frequently fails. 

However, right now if `stress_max_iteration_number` is reached there is no message to indicate the strain rate residual threshold was not reached.

naliboff commented on tests/visco_plastic_peierls_mei_exact.prm:
```diff
@@ -0,0 +1,115 @@
+    set Stress exponents for Peierls creep        = 2.0
+    set Peierls fitting parameters                = 0.15
+    set Peierls glide parameters p                = 0.5
+    set Peierls glide parameters q                = 1.0
```
Pleased add in the following lines as a reminder for those looking at the test that these are additional options with the exact method:
`set Peierls strain rate residual tolerance = 1e-22`
`set Maximum Peierls strain rate iterations = 40`

naliboff commented:
/rebuild

bobmyhill commented:
Hi @naliboff,
I've addressed your comments, and added an AssertThrow to make sure that ASPECT terminates if the Newton iterative scheme fails.

There are a few changes that will probably need to be made to this rheology when it starts being used in models. Yes, we may need a backup iterative scheme (although I'm hopeful that we won't need this as we're using Magali's approximate scheme as a starting point). There should also probably be a limiter on the min and max viscosities for both the approximate and exact flow laws. 

I'd argue that we should save these modifications until we know we have a problem. I suspect that the problems will make themselves apparent very quickly in "realistic" models using only Peierls creep.
